<!DOCTYPE html><html><head><meta charset="utf-8"><meta name="X-UA-Compatible" content="IE=edge"><title> RabbitMQ报错[ERROR: node with name 'rabbit' already running on 'localhost'] · Liqun's Homepage</title><meta name="description" content="RabbitMQ报错[ERROR: node with name 'rabbit' already running on 'localhost'] - Liqun Li"><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="icon" href="/blog/favicon2.jpeg"><link rel="stylesheet" href="/blog/css/apollo.css"><link rel="search" type="application/opensearchdescription+xml" href="liqul.github.io/blog/atom.xml" title="Liqun's Homepage"></head><body><div class="wrap"><header><a href="/blog/" class="logo-link"><img src="/blog/favicon2.jpeg" alt="logo"></a><ul class="nav nav-list"><li class="nav-list-item"><a href="/blog/" target="_self" class="nav-list-link">BLOG</a></li><li class="nav-list-item"><a href="/blog/archives/" target="_self" class="nav-list-link">ARCHIVE</a></li><li class="nav-list-item"><a href="/blog/about/" target="_self" class="nav-list-link">ABOUT</a></li><li class="nav-list-item"><a href="https://scholar.google.com/citations?user=icgetesAAAAJ&amp;hl=en" target="_blank" class="nav-list-link">GOOGLE SCHOLAR</a></li><li class="nav-list-item"><a href="/blog/atom.xml" target="_self" class="nav-list-link">RSS</a></li></ul></header><main class="container"><div class="post"><article class="post-block"><h1 class="post-title">RabbitMQ报错[ERROR: node with name 'rabbit' already running on 'localhost']</h1><div class="post-info">Sep 7, 2015</div><div class="post-content"><p>这个报错是在执行./rabbitmq-server -detached的时候报出来的，除了这句以外没有任何信息。这个报错的本意是rabbit@localhost已经在运行了，不允许再启动一个实例。于是尝试干掉已有的进程：<br><!--break--></p>
<p><pre class="lang:default decode:true ">ps aux | grep rabbit</pre><br>执行后发现没有任何进程。尝试Google类似的问题，找了很久终于发现一个可能的<a href="http://stackoverflow.com/questions/8737754/node-with-name-rabbit-already-running-but-also-unable-to-connect-to-node" target="_blank" rel="noopener">问题</a>——/etc/hosts文件可能被改出问题了。</p>
<p>这要回到RabbitMQ的一个基本认识上，它是通过hostname来唯一确定本机的，甚至包括队列存储都与hostname直接对应，如果hostname被改了，意味着实例本身都变化了。于是执行</p>
<p><pre class="lang:default decode:true ">hostname -s</pre><br>发现居然没问题，还是localhost。所以，的确感觉非常奇怪。</p>
<p>进一步检查/etc/hosts文件，发现无意间把localhost指向了局域网内的另一台机器：</p>
<p><pre class="lang:default decode:true ">192.168.1.56 localhost</pre><br>于是终于找到原因了，RabbitMQ启动时检查localhost，发现上面已经启动了一个RabbitMQ实例，于是报了最前面的错误。只需要去掉hosts文件里的配置就好了。</p>
<p>上面说的是主要原因，另一个意外操作导致这个错误的原因是erlang的守护进程被其它用户启动了，这可以通过：</p>
<p><pre class="lang:default decode:true ">ps aux | grep epmd</pre><br>检查。需要干掉这个进程才能顺利把RabbitMQ启动起来。</p>
</div></article></div></main><footer><div class="paginator"><a href="/blog/imu/" class="prev">PREV</a><a href="/blog/dependency-injection/" class="next">NEXT</a></div><div class="copyright"><p>© 2016 - 2018 <a href="liqul.github.io/blog">Liqun Li</a>, powered by <a href="https://hexo.io/" target="_blank">Hexo</a> and <a href="https://github.com/pinggod/hexo-theme-apollo" target="_blank">hexo-theme-apollo</a>.</p></div></footer></div><script async src="//cdn.bootcss.com/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" integrity="sha384-crwIf/BuaWM9rM65iM+dWFldgQ1Un8jWZMuh3puxb8TOY9+linwLoI7ZHZT+aekW" crossorigin="anonymous"></script></body></html>